common/lib/transport/comettransport.ts

```diff
@@ -1,165 +1,177 @@
+
+const REQ_SEND = 0,
+	REQ_RECV = 1,
+	REQ_RECV_POLL = 2,
+	REQ_RECV_STREAM = 3;
+
+/* TODO: can remove once realtime sends protocol message responses for comet errors */
+function shouldBeErrorAction(err: ErrorInfo) {
+	const UNRESOLVABLE_ERROR_CODES = [80015, 80017, 80030];
+	if(err.code) {
+		if(Auth.isTokenErr(err)) return false;
+		if(Utils.arrIn(UNRESOLVABLE_ERROR_CODES, err.code)) return true;
+		return (err.code >= 40000 && err.code < 50000);
+	} else {
+		/* Likely a network or transport error of some kind. Certainly not fatal to the connection */
+		return false;
 	}
-
-	function protocolMessageFromRawError(err) {
-		/* err will be either a legacy (non-protocolmessage) comet error response
-		 * (which will have an err.code), or a xhr/network error (which won't). */
-		if(shouldBeErrorAction(err)) {
-			return [ProtocolMessage.fromValues({action: ProtocolMessage.Action.ERROR, error: err})];
-		} else {
-			return [ProtocolMessage.fromValues({action: ProtocolMessage.Action.DISCONNECTED, error: err})];
-		}
+}
+
+function protocolMessageFromRawError(err: ErrorInfo) {
+	/* err will be either a legacy (non-protocolmessage) comet error response
+		* (which will have an err.code), or a xhr/network error (which won't). */
+	if(shouldBeErrorAction(err)) {
+		return [ProtocolMessage.fromValues({action: ProtocolMessage.Action.ERROR, error: err})];
+	} else {
+		return [ProtocolMessage.fromValues({action: ProtocolMessage.Action.DISCONNECTED, error: err})];
 	}
+}
 
 	/*
 	 * A base comet transport class
 	 */
-	function CometTransport(connectionManager, auth, params) {
-		/* binary not supported for comet, so just fall back to default */
-		params.format = undefined;
-		params.heartbeats = true;
-		Transport.call(this, connectionManager, auth, params);
-		/* streaming defaults to true */
+abstract class CometTransport extends Transport {
+	stream: string | boolean;
+	sendRequest: IXHRRequest | null;
+	recvRequest: null | IXHRRequest;
+	pendingCallback: null;
+	pendingItems: null | Array<ProtocolMessage>;
+	baseUri?: string;
+	authParams?: Record<string, any>;
+	closeUri?: string;
+	disconnectUri?: string;
+	sendUri?: string;
+	recvUri?: string;
+
+	constructor(connectionManager: ConnectionManager, auth: Auth, params: TransportParams) {
+		super(connectionManager, auth, params, /* binary not supported for comet so force JSON protocol */ true);
 		this.stream = ('stream' in params) ? params.stream : true;
 		this.sendRequest = null;
 		this.recvRequest = null;
 		this.pendingCallback = null;
 		this.pendingItems = null;
 	}
-	Utils.inherits(CometTransport, Transport);
 
-	CometTransport.REQ_SEND = REQ_SEND;
-	CometTransport.REQ_RECV = REQ_RECV;
-	CometTransport.REQ_RECV_POLL = REQ_RECV_POLL;
-	CometTransport.REQ_RECV_STREAM = REQ_RECV_STREAM;
+	abstract createRequest(uri: string, headers: Record<string, string> | null, params?: Record<string, unknown> | null, body?: unknown, requestMode?: number): IXHRRequest;
+
+	static REQ_SEND = REQ_SEND;
```

Comment:
Curious here around use of `static`, relating to [previous comment](https://github.com/ably/ably-js/pull/834#discussion_r736296323).

Reply:
As in `protocol.js`, this is an artefact of the way the library was modularised before webpack (basically you could only export one thing from a module so adding a static property was a hacky way to add a named export to the module alongside the default export). Fortunately, by creating the XHRStates enum in e9d0e80 this has already been taken care of.